iree/compiler/Dialect/Flow/Transforms/ConvertLinalgMatmulToMmt4D.cpp

```diff
@@ -70,19 +91,125 @@ static Value transpose(mlir::Location loc, PatternRewriter &rewriter,
   return transposedOp.getResult(0);
 };
 
-// Collapses a 4d tensor operand to 2d given its target shape.
+// Collapses a 4d tensor input to 2d given its target shape.
 // Example: (M1, M0, N1, N0) -> (M, N)
 static Value collapseTo2D(mlir::Location loc, PatternRewriter &rewriter,
-                          Value operand, ArrayRef<int64_t> targetShape) {
-  auto operandType = operand.getType().cast<RankedTensorType>();
+                          Value input, ArrayRef<int64_t> targetShape) {
+  auto inputType = input.getType().cast<RankedTensorType>();
   auto targetType =
-      RankedTensorType::get(targetShape, operandType.getElementType());
+      RankedTensorType::get(targetShape, inputType.getElementType());
   SmallVector<ReassociationIndices> collapseIndices = {{0, 1}, {2, 3}};
   Value reshapedOperand = rewriter.create<linalg::TensorCollapseShapeOp>(
-      loc, targetType, operand, collapseIndices);
+      loc, targetType, input, collapseIndices);
   return reshapedOperand;
 }
 
+// Returns true if an input of the given |inputStaticShape| needs padding to
+// ensure that its shape will be a multiple of |tileShape|. That's always true
+// in the dynamic shape case.
+static bool needsPadding(ArrayRef<int64_t> inputStaticShape,
+                         ArrayRef<int64_t> tileShape) {
+  assert(inputStaticShape.size() == tileShape.size());
+  for (int i = 0; i < inputStaticShape.size(); i++) {
+    if (inputStaticShape[i] == ShapedType::kDynamicSize) {
+      return true;
+    }
+    if (inputStaticShape[i] % tileShape[i]) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Pads |input| on the bottom and on the right to the next multiple of
+// |tileShape|.
+static Value pad(Location loc, PatternRewriter &rewriter, Value input,
+                 ArrayRef<int64_t> tileShape) {
+  SmallVector<OpFoldResult, 2> lowPadding, highPadding;
+  SmallVector<int64_t, 2> resultTypeShape;
```

Comment:
Do you have good reason to believe these are the right values here but also *don't* know that there's a fixed upper bound? I kind of suspect it's one or the other in this case, so you could drop the explicit size

> In the absence of a well-motivated choice for the number of inlined elements N, it is recommended to use SmallVector<T> (that is, omitting the N). This will choose a default number of inlined elements reasonable for allocation on the stack (for example, trying to keep sizeof(SmallVector<T>) around 64 bytes).

https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

or use an array: https://llvm.org/docs/ProgrammersManual.html#fixed-size-arrays

Reply:
reasonable reason: the value here will be N when this pad function is called on a N-D tensor. In practice we call it on 2-D tensors, so N=2.

We could hardcode the 2D case here and in other functions in this file. I was just keeping the existing style in this file, of not hardcoding the dimensionality in functions that readily apply to any dimensionality. We could do that, we certainly would if any of this matters for performance.

Reply:
Keeping it general SGTM. In that case I would probably drop the explicit N though. The inference logic is pretty good

Reply:
i do agree that it's probably futile to specify the N, but then more generally i suspect that all of the performance considerations here are futile because all of the code we're looking at here is very "dynamic" ie slow code anyway... every thing that we do in these transformation involves long chains of dereferences, this code is not the kind of "statically typed" c++ code where having the right vector container and avoiding small heap allocations would be the bottleneck...  i feel that i'm missing a big piece of context in all this llvm code.

Reply:
I don't think this is so much about optimizing performance as about avoiding pessimizing performance while making code less readable. When I see an explicit N I try to understand why. It's kind of like having a std::move on a return value: unless you really think you know better than the compiler in this case, leave it off.

My comments on using std::array are similar (though I feel less strongly). When I see `SmallVector<int> foo = {1, 2}` I look for the place where you're mutating the SmallVector.